cpp/src/arrow/dataset/dataset.h

```diff
@@ -81,9 +81,32 @@ class ARROW_DS_EXPORT DataSource {
   virtual std::unique_ptr<DataFragmentIterator> GetFragments(
       std::shared_ptr<ScanOptions> options) = 0;
 
+  /// \brief An expression which evaluates to true for all data viewed by this DataSource.
+  /// May be null, which indicates no information is available.
+  const std::shared_ptr<Expression>& partition_expression() const {
+    return partition_expression_;
+  }
+
+  /// FIXME(bkietz) providing a simple mutator like this is probably not ideal
+  void partition_expression(std::shared_ptr<Expression> e) {
+    partition_expression_ = std::move(e);
+  }
+
   virtual std::string type() const = 0;
 
   virtual ~DataSource() = default;
+
+ protected:
+  DataSource() = default;
+  explicit DataSource(std::shared_ptr<Expression> c)
+      : partition_expression_(std::move(c)) {}
+
+  /// Mutates a ScanOptions by assuming condition_ holds for all yielded fragments.
+  /// Returns non-null if context->selector is not satisfiable in this DataSource.
+  std::unique_ptr<EmptyIterator<std::shared_ptr<DataFragment>>> AssumeCondition(
```

Comment:
Multiples questions: 
- Did you mean to return `Iterator<DataFragment>` (why empty)?
- I'd reverse the API with inherited classes, i.e. DataFragments() is implemented in DataSource (and skips accordingly) or calls the virtual `FindABetterNameDataFragments()`.
- Make a protected/private function `bool PassPredicate(options)`.

Reply:
- This method will only return null or an emtpy iterator if the selector contradicts the partition_expression. Since it only returns empty or null, the return type can be unique_ptr<EmptyIterator>. I could refactor it to return false/true if you prefer. Is this what you mean with PassPredicate?
- Not sure what you mean
-

Reply:
I think I see what you mean. I've made DataSource::GetFragments() public non virtual, DataSource::GetFragmentsImpl() protected virtual. GetFragments() returns an empty iterator if the selector contradicts the partition_expression and defers to GetFragmentsImpl otherwise.